src/Search/Services/SearchableService.php

```diff
@@ -0,0 +1,162 @@
+    {
+        $this->cache = [];
+    }
+
+    /**
+     * Used during search reindex
+     *
+     * This is considered the primary layer of protection
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isIndexable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, true);
+    }
+
+    /**
+     * Used when retrieving search results
+     *
+     * This is considered the secondary layer of protection
+     *
+     * It's important to still have this layer in conjuction with the index layer as non-searchable results may be
+     * in the search index because:
+     * a) they were added to the index pre-fulltextsearch 3.7 and a reindex to purge old records was never run, OR
+     * b) the DataObject has a non-deterministic canView() check such as `return $date <= $dateOfIndex;`
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isViewable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, false);
+    }
+
+    /**
+     * Checks and caches whether the given DataObject can be indexed. This is determined by two factors:
+     * - Whether the ShowInSearch property / getShowInSearch() method evaluates to true
+     * - Whether the canView method evaluates to true against an anonymous user (optional, can be disabled)
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    private function isSearchable(DataObject $obj, bool $indexing): bool
+    {
+        // check if is a valid DataObject that has been persisted to the database
+        if (is_null($obj) || !$obj->ID) {
+            return false;
+        }
+
+        $key = $this->getCacheKey($obj);
+        if (isset($this->cache[$key])) {
+            return $this->cache[$key];
+        }
+
+        $value = true;
+
+        // ShowInSearch check
+        // This will also call $obj->getShowInSearch() if it exists
+        if (isset($obj->ShowInSearch) && !$obj->ShowInSearch) {
+            $value = false;
+        }
+
+        // Anonymous member canView()
+        if ($value) {
+            $objClass = $obj->getClassName();
+            if ($indexing) {
+                if (!$this->classSkipsCanViewCheck($objClass)) {
+                    $value = Member::actAs(null, function () use ($obj) {
+                        return $obj->canView();
+                    });
+                }
+            } else {
+                $value = $obj->canView();
+            }
+        }
+        $this->extend('updateIsSearchable', $obj, $indexing, $value);
+        $this->cache[$key] = $value;
+        return $value;
+    }
+
+    /**
+     * Generates a unique key to cache each DataObject with. Can be extended via updateCacheKey.
+     *
+     * @param DataObject $obj
+     * @return string
+     */
+    private function getCacheKey(DataObject $obj): string
+    {
+        $key = $obj->ClassName . '_' . $obj->ID;
+        $this->extend('updateCacheKey', $obj, $key);
```

Comment:
I'm not personally familiar with the data structure that fluent uses for different language variants.  Is the english and french version of the 'Content' field effectively stored in different database columns on the same table? (so $obj->ClassName and $obj->ID are the same?)

Reply:
Have switched this to use the freshly merged `$obj->getUniqueKey()`, which itself utilises `cacheKeyComponent` so I believe fluent will be covered by that?

https://github.com/silverstripe/silverstripe-framework/pull/9400/files

Have left the `updateCacheKey` extension here in case it's needed for some other unforeseen use-case, though my assumption here is that fluent/subsites/versioned/other will all be covered in `$obj->getUniqueKey()`

Is this a fair assumption to make?

Reply:
Yes that's a fair assumption. Great to see this PR in action already, thanks